community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java

```diff
@@ -242,139 +315,176 @@ private PagedFile openOrCreate( PageCache pageCache, File indexFile,
+    {
+        try ( PageCursor metaCursor = openMetaPageCursor( pagedFile ) )
+        {
+            metaCursor.putInt( pageSize );
+            metaCursor.putLong( layout.identifier() );
+            metaCursor.putInt( layout.majorVersion() );
+            metaCursor.putInt( layout.minorVersion() );
+            layout.writeMetaData( metaCursor );
+            checkOutOfBounds( metaCursor );
+        }
+    }
 
-                bTreeNode.goTo( cursor, childId, stableGeneration, unstableGeneration );
+    private PagedFile mapWithCorrectPageSize( PageCache pageCache, File indexFile, PagedFile pagedFile )
+            throws IOException
+    {
+        // This index was created with another page size, re-open with that actual page size
+        if ( pageSize != pageCache.pageSize() )
+        {
+            if ( pageSize > pageCache.pageSize() )
+            {
+                throw new MetadataMismatchException( "Tree in " + indexFile.getAbsolutePath() +
+                        " was created with page size:" + pageSize +
+                        ", but page cache used to open it this time has a smaller page size:" +
+                        pageCache.pageSize() + " so cannot be opened" );
             }
+            pagedFile.close();
+            return pageCache.map( indexFile, pageSize );
         }
-        while ( isInternal && keyCount > 0 );
+        return pagedFile;
+    }
+
+    @Override
+    public RawCursor<Hit<KEY,VALUE>,IOException> seek( KEY fromInclusive, KEY toExclusive ) throws IOException
+    {
+        PageCursor cursor = pagedFile.io( rootId, PagedFile.PF_SHARED_READ_LOCK );
+        KEY key = layout.newKey();
+        VALUE value = layout.newValue();
+        long generation = this.generation;
+        long stableGeneration = stableGeneration( generation );
+        long unstableGeneration = unstableGeneration( generation );
+        goToRoot( cursor, stableGeneration, unstableGeneration );
 
         // Returns cursor which is now initiated with left-most leaf node for the specified range
         return new SeekCursor<>( cursor, key, value, bTreeNode, fromInclusive, toExclusive, layout,
-                stableGeneration, unstableGeneration, pos, keyCount );
+                stableGeneration, unstableGeneration, generationSupplier );
     }
 
     @Override
-    public long acquireNewId()
+    public long acquireNewId() throws IOException
     {
         lastId++;
-        return lastId;
-    }
-
-    // Utility method
-    public void printTree() throws IOException
-    {
-        try ( PageCursor cursor = pagedFile.io( rootId, PagedFile.PF_SHARED_READ_LOCK ) )
+        try ( PageCursor cursor = pagedFile.io( lastId, PagedFile.PF_SHARED_WRITE_LOCK ) )
         {
             cursor.next();
-            TreePrinter.printTree( cursor, bTreeNode, layout, stableGeneration, unstableGeneration, System.out );
+            // TODO use cursor.clear() when available
```

Comment:
This code has moved as part of free-list work and changed to use `zapPage()` so will ignore in this PR.